internal/cmd/gocdk/fswatch.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2019 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"golang.org/x/xerrors"
+)
+
+// waitForDirChange blocks until the directory changes or the context's Done
+// channel is closed.
+func waitForDirChange(ctx context.Context, path string, pollInterval time.Duration) error {
+	tick := time.NewTicker(pollInterval)
+	defer tick.Stop()
+	init, err := snapshotTree(path)
+	if err != nil {
+		return xerrors.Errorf("wait for %s to change: %w", path, err)
+	}
+	for {
+		select {
+		case <-tick.C:
+			same, err := init.sameAs(path)
```

Comment:
WDYT about doing an explicit snapshot here and calling `recursiveEqual` (and dropping `sameAs`)?

Reply:
I left a TODO in `sameAs`: in the case where there has been a change, `sameAs` could use less stat calls than a full snapshot because it could find the difference and return early before descending on the walk. However, I recognize that this only optimizes for the edge event, not the steady-state. WDYT?